shell/server/pre-meteor.js

```diff
@@ -341,180 +341,206 @@ function serveStaticAsset(req, res) {
   });
 }
 
-Meteor.startup(() => {
+const canonicalizeShellOrWildcardUrl = (hostname, url) => {
+  // Start with ROOT_URL, apply host & path from inbound URL, then
+  // redirect.
+  let targetUrl = Url.parse(process.env.ROOT_URL);
+
+  // Retain the protocol & port from ROOT_URL but use the inbound
+  // hostname.
+  targetUrl.host = hostname + ':' + targetUrl.port;
+
+  // The following allows to avoid decoding + re-encoding query
+  // string parameters, if provided.
+  targetUrl = Url.resolve(Url.format(targetUrl), url);
+  return targetUrl;
+};
+
+const dispatchToMeteorOrStaticPublishing = (req, res, next, redirectRatherThanServeShell) => {
+  if (!req.headers.host) {
```

Comment:
It is still needed for the `redirectToMeteorOrServeStaticPublishing` codepath (though perhaps that suggests that the check should be moved out to there).